File: android/app/src/main/kotlin/com/tetherhealthgroup/tetherhealth/unplug/InterceptTokens.kt
package com.tetherhealthgroup.tetherhealth.unplug

import android.content.Context
import android.graphics.Color
import io.flutter.FlutterInjector
import org.json.JSONObject

/**
 * The Kotlin reader for the shared intercept tokens.
 *
 * Addendum §2.1 warns that the intercept, built three times, drifts apart within
 * two sprints unless the colours and copy live in one file all three read. That
 * file is `assets/unplug/intercept_tokens.json`, shipped in the Flutter asset
 * bundle; this is the Android reader for it, and the overlay draws nothing that
 * is not in here.
 *
 * Parsing is strict, exactly as the Dart reader is. A malformed token file is a
 * file the other two implementations will also fail on, and quietly falling back
 * to a hard-coded colour would hide precisely the drift this file prevents.
 */
data class InterceptTokens(
    val version: String,
    val background: Int,
    val surface: Int,
    val onBackground: Int,
    val onSurfaceMuted: Int,
    val accent: Int,
    val onAccent: Int,
    val danger: Int,
    val title: String,
    val body: String,
    val reasonLabel: String,
    val breatheAction: String,
    val closeAction: String,
    val overrideAction: String,
    val overrideRemaining: String,
    val overrideExhausted: String,
    val strictNotice: String,
    val breathSeconds: Int,
    val latencyMinMs: Int,
    val latencyMaxMs: Int
) {

    /** Substitutes `{name}` placeholders, matching the Dart reader's contract. */
    fun format(template: String, values: Map<String, String>): String =
        values.entries.fold(template) { text, entry ->
            text.replace("{${entry.key}}", entry.value)
        }

    companion object {
        const val ASSET_PATH = "assets/unplug/intercept_tokens.json"

        @Volatile
        private var cached: InterceptTokens? = null

        /**
         * Reads and caches the token file.
         *
         * @throws IllegalStateException when the file is missing or malformed.
         */
        fun load(context: Context): InterceptTokens {
            cached?.let { return it }
            val key = FlutterInjector.instance()
                .flutterLoader()
                .getLookupKeyForAsset(ASSET_PATH)
            val raw = try {
                context.assets.open(key).bufferedReader().use { it.readText() }
            } catch (error: Exception) {
                throw IllegalStateException(
                    "The shared intercept token file could not be read at $key. " +
                        "The Dart and SwiftUI intercepts read the same file.",
                    error
                )
            }
            return parse(raw).also { cached = it }
        }

        fun parse(raw: String): InterceptTokens {
            val root = JSONObject(raw)
            val colors = root.getJSONObject("colors")
            val copy = root.getJSONObject("copy")
            val timing = root.getJSONObject("timing")
            return InterceptTokens(
                version = root.getString("version"),
                background = color(colors, "background"),
                surface = color(colors, "surface"),
                onBackground = color(colors, "onBackground"),
                onSurfaceMuted = color(colors, "onSurfaceMuted"),
                accent = color(colors, "accent"),
                onAccent = color(colors, "onAccent"),
                danger = color(colors, "danger"),
                title = copy.getString("title"),
                body = copy.getString("body"),
                reasonLabel = copy.getString("reasonLabel"),
                breatheAction = copy.getString("breatheAction"),
                closeAction = copy.getString("closeAction"),
                overrideAction = copy.getString("overrideAction"),
                overrideRemaining = copy.getString("overrideRemaining"),
                overrideExhausted = copy.getString("overrideExhausted"),
                strictNotice = copy.getString("strictNotice"),
                breathSeconds = timing.getInt("breathSeconds"),
                latencyMinMs = timing.getInt("androidOverlayLatencyMsMin"),
                latencyMaxMs = timing.getInt("androidOverlayLatencyMsMax")
            )
        }

        /** Parses `#RRGGBB`, the form the Dart and Swift readers also accept. */
        private fun color(source: JSONObject, name: String): Int {
            val value = source.getString(name)
            require(value.length == 7 && value.startsWith("#")) {
                "Intercept colour \"$name\" is not #RRGGBB."
            }
            return Color.parseColor(value)
        }
    }
}

File: lib/unplug/models/intercept_tokens.dart
import 'dart:convert';
import 'dart:ui' show Color;

import 'package:flutter/services.dart' show rootBundle;

/// The design tokens and copy shared by all three intercept implementations.
///
/// Addendum §2.1 requires the intercept to be built three times — Dart, SwiftUI
/// and an Android overlay — and warns that the three drift apart within two
/// sprints unless the colours and copy live in one file all of them read. That
/// file is `assets/unplug/intercept_tokens.json`; this class is the Dart reader
/// for it.
///
/// Parsing is strict on purpose. A missing key is a token file the native
/// implementations will also fail on, and silently substituting a default here
/// would hide exactly the drift this file exists to prevent.
class InterceptTokens {
  const InterceptTokens({
    required this.version,
    required this.colors,
    required this.copy,
    required this.breathSeconds,
    required this.androidLatencyMsMin,
    required this.androidLatencyMsMax,
  });

  /// Where the token file lives in the asset bundle.
  static const assetPath = 'assets/unplug/intercept_tokens.json';

  final String version;
  final Map<String, Color> colors;
  final Map<String, String> copy;
  final int breathSeconds;
  final int androidLatencyMsMin;
  final int androidLatencyMsMax;

  Color color(String name) {
    final value = colors[name];
    if (value == null) {
      throw StateError('Intercept token file has no colour named "$name".');
    }
    return value;
  }

  /// Returns a copy string, substituting `{name}` placeholders.
  String text(String name, [Map<String, String> values = const {}]) {
    final template = copy[name];
    if (template == null) {
      throw StateError(
          'Intercept token file has no copy string named "$name".');
    }
    return values.entries.fold(
      template,
      (result, entry) => result.replaceAll('{${entry.key}}', entry.value),
    );
  }

  static Future<InterceptTokens> load() async {
    return parse(await rootBundle.loadString(assetPath));
  }

  static InterceptTokens parse(String source) {
    final decoded = jsonDecode(source);
    if (decoded is! Map<String, dynamic>) {
      throw const FormatException('Intercept token file is not a JSON object.');
    }

    final colors = _section(decoded, 'colors');
    final copy = _section(decoded, 'copy');
    final timing = _section(decoded, 'timing');

    return InterceptTokens(
      version: _string(decoded, 'version'),
      colors: {
        for (final entry in colors.entries)
          entry.key: _color(entry.key, entry.value),
      },
      copy: {
        for (final entry in copy.entries) entry.key: _copyString(entry),
      },
      breathSeconds: _int(timing, 'breathSeconds'),
      androidLatencyMsMin: _int(timing, 'androidOverlayLatencyMsMin'),
      androidLatencyMsMax: _int(timing, 'androidOverlayLatencyMsMax'),
    );
  }

  static Map<String, dynamic> _section(Map<String, dynamic> root, String key) {
    final value = root[key];
    if (value is! Map<String, dynamic>) {
      throw FormatException('Intercept token file has no "$key" object.');
    }
    return value;
  }

  static String _string(Map<String, dynamic> root, String key) {
    final value = root[key];
    if (value is! String) {
      throw FormatException('Intercept token "$key" is not a string.');
    }
    return value;
  }

  static String _copyString(MapEntry<String, dynamic> entry) {
    final value = entry.value;
    if (value is! String) {
      throw FormatException('Intercept copy "${entry.key}" is not a string.');
    }
    return value;
  }

  static int _int(Map<String, dynamic> root, String key) {
    final value = root[key];
    if (value is! int) {
      throw FormatException('Intercept token "$key" is not an integer.');
    }
    return value;
  }

  /// Parses `#RRGGBB`, the form the native implementations also read.
  static Color _color(String name, Object? value) {
    if (value is! String || value.length != 7 || !value.startsWith('#')) {
      throw FormatException('Intercept colour "$name" is not #RRGGBB.');
    }
    final channels = int.tryParse(value.substring(1), radix: 16);
    if (channels == null) {
      throw FormatException('Intercept colour "$name" is not hexadecimal.');
    }
    return Color(0xFF000000 | channels);
  }
}
